### Загрузка датасета из kaggle.com

In [1]:
!gdown --id 1XQWW7lsK0EiaibyDnSfQ1wKoZqdDZ2xu
! mkdir ~/.kaggle
! cp kaggle.json ~/.kaggle/
! chmod 600 ~/.kaggle/kaggle.json
! kaggle datasets download -d yasserh/titanic-dataset
! unzip titanic-dataset.zip -d titanic-dataset

/usr/local/lib/python3.10/dist-packages/gdown/cli.py:121: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=1XQWW7lsK0EiaibyDnSfQ1wKoZqdDZ2xu
To: /content/kaggle.json
100% 73.0/73.0 [00:00<00:00, 275kB/s]
  0% 0.00/22.0k [00:00<?, ?B/s]
100% 22.0k/22.0k [00:00<00:00, 9.12MB/s]
Archive:  titanic-dataset.zip
  inflating: titanic-dataset/Titanic-Dataset.csv  


### Импорт библиотек и работа с датасетом

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('/content/titanic-dataset/Titanic-Dataset.csv')

In [4]:
df.sample(5)

PassengerId  Survived  Pclass                            Name     Sex  \
45            46         0       3        Rogers, Mr. William John    male   
241          242         1       3  Murphy, Miss. Katherine "Kate"  female   
868          869         0       3     van Melkebeke, Mr. Philemon    male   
245          246         0       1     Minahan, Dr. William Edward    male   
632          633         1       1       Stahelin-Maeglin, Dr. Max    male   

      Age  SibSp  Parch           Ticket   Fare Cabin Embarked  
45    NaN      0      0  S.C./A.4. 23567   8.05   NaN        S  
241   NaN      1      0           367230  15.50   NaN        Q  
868   NaN      0      0           345777   9.50   NaN        S  
245  44.0      2      0            19928  90.00   C78        Q  
632  32.0      0      0            13214  30.50   B50        C

#### Определить количество пустых ячеек

In [5]:
# Количество пустых ячеек в каждом столбце
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [6]:
# Общее количество пустых ячеек
print(f'Общее количество пустых ячеек = {df.isnull().sum().sum()}')

Общее количество пустых ячеек = 866


#### Заполнение пустых ячеек

In [7]:
# В колонке Age, 177 пропущенных значений
# Зполним пропуски средним значением
df['Age'].fillna(df['Age'].mean(), inplace = True)

In [8]:
df.sample(5)

PassengerId  Survived  Pclass                           Name     Sex  \
615          616         1       2            Herman, Miss. Alice  female   
470          471         0       3              Keefe, Mr. Arthur    male   
318          319         1       1       Wick, Miss. Mary Natalie  female   
87            88         0       3  Slocovski, Mr. Selman Francis    male   
73            74         0       3    Chronopoulos, Mr. Apostolos    male   

           Age  SibSp  Parch           Ticket      Fare Cabin Embarked  
615  24.000000      1      2           220845   65.0000   NaN        S  
470  29.699118      0      0           323592    7.2500   NaN        S  
318  31.000000      0      2            36928  164.8667    C7        S  
87   29.699118      0      0  SOTON/OQ 392086    8.0500   NaN        S  
73   26.000000      1      0             2680   14.4542   NaN        C

In [9]:
# В колонке Cabin больше всего пропущенных значений
# Проверим существует ли какая-то взаимосвязь между выживанием и тем, была ли у пассажира отдельная каюта
df['Survived'].groupby(df['Cabin'].isnull()).mean()

Cabin
False    0.666667
True     0.299854
Name: Survived, dtype: float64

 Пассажиры у кого запись отсутствует - выжили около 30%. А у кого запись о наличии каюты есть - выжило 67%, поэтому заменим значения в колонке Cabin категориальным признаком (0 - нет отдельной каюты, 1 - есть каюта)




In [10]:
df['Cabin'] = np.where(df['Cabin'].isnull(), 0,1)

In [11]:
df.sample(5)

PassengerId  Survived  Pclass                             Name   Sex  \
664          665         1       3      Lindqvist, Mr. Eino William  male   
386          387         0       3  Goodwin, Master. Sidney Leonard  male   
524          525         0       3                Kassem, Mr. Fared  male   
622          623         1       3                 Nakid, Mr. Sahid  male   
243          244         0       3    Maenpaa, Mr. Matti Alexanteri  male   

           Age  SibSp  Parch             Ticket     Fare  Cabin Embarked  
664  20.000000      1      0  STON/O 2. 3101285   7.9250      0        S  
386   1.000000      5      2            CA 2144  46.9000      0        S  
524  29.699118      0      0               2700   7.2292      0        C  
622  20.000000      1      1               2653  15.7417      0        C  
243  22.000000      0      0  STON/O 2. 3101275   7.1250      0        S

В колонке Embarked два пропущенных значения, заполним их наиболее часто встречаемым значением

In [12]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

Наиболее часто встречаемое значение "S"

In [13]:
df['Embarked'].replace(np.nan, 'S', inplace = True)

#### Аналитика

1. Количество уникальных значений

In [14]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             89
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin            2
Embarked         3
dtype: int64

2. Минимальное и максимальное значение

In [15]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   13.002015    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   29.699118    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare       Cabin  
count  891.000000  891.000000  891.000000  
mean     0.381594   32.204208    0.228956  
std      0.806057   49.693429    0.420397  
min      0.000000    0.000000    0.000000  
25%      0.000000    7.910400    0.000000  
50%      0.000000   14.454200    0.000000  
75%      0.000000   31.000000    0.000000  
max      6.000000  512.329200    1.000000

In [16]:
df.agg(['min','max'])

PassengerId  Survived  Pclass                         Name     Sex  \
min            1         0       1          Abbing, Mr. Anthony  female   
max          891         1       3  van Melkebeke, Mr. Philemon    male   

       Age  SibSp  Parch     Ticket      Fare  Cabin Embarked  
min   0.42      0      0     110152    0.0000      0        C  
max  80.00      8      6  WE/P 5735  512.3292      1        S



*   Количество пассажиров = 891
*   Всего 3 класса круиза
*   Возраст пассажиров от меньше 1 года до 80
*   Максимальное число братьев,сестер или супругов на борту у человека = 8
*   Максимальное число людей в семье (родители, дети) = 6
*   Максимальная цена билета = $512
*   Всего 3 порта посадки










3. Влияние всех признаков напризнак Survived

In [17]:
df['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Из 891 пассажиров выжили 342

In [19]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare  Cabin Embarked  
0        1      0         A/5 21171   7.2500      0        S  
1        1      0          PC 17599  71.2833      1        C  
2        0      0  STON/O2. 3101282   7.9250      0        S  
3        1      0            113803  53.1000      1        S  
4        0      0            373450   8.0500      0        S  
..     ...    ...               ...      ...    ...      ...  
886      0      0            211536  13.0000      0        S  
887      0      0            112053  30.0000      1        S  
888      1      2        W./C. 6607  23.4500      0        S  
889      0      0            111369  30.0000      1        C  
890      0      0            370376   7.7500      0        Q  

[891 rows x 12 columns]

In [20]:
surv = df.groupby(df['Name'].str.split(r'[,.]').str[1])
surv['Survived'].mean()

Name
 Capt            0.000000
 Col             0.500000
 Don             0.000000
 Dr              0.428571
 Jonkheer        0.000000
 Lady            1.000000
 Major           0.500000
 Master          0.575000
 Miss            0.697802
 Mlle            1.000000
 Mme             1.000000
 Mr              0.156673
 Mrs             0.792000
 Ms              1.000000
 Rev             0.000000
 Sir             1.000000
 the Countess    1.000000
Name: Survived, dtype: float64

Среднее значение выживших в зависимости от приставки в имени. Например Sir, Lady люди из высокого класса выжили в 100%

In [21]:
df.groupby(['Pclass','Sex'])['Survived'].mean().unstack(level=1)

Sex       female      male
Pclass                    
1       0.968085  0.368852
2       0.921053  0.157407
3       0.500000  0.135447

В каждом классе доля выживших женщин больше чем мужчин

In [22]:
# Средний возраст выживших в каждом классе
s = df[df['Survived'] == 1]
pd.pivot_table(s, values='Age', index='Pclass', columns='Sex', aggfunc='mean').round()

Sex     female  male
Pclass              
1         34.0  36.0
2         28.0  18.0
3         23.0  24.0

In [23]:
pd.pivot_table(df, index='SibSp', columns='Survived', values='PassengerId', aggfunc='count')

Survived      0      1
SibSp                 
0         398.0  210.0
1          97.0  112.0
2          15.0   13.0
3          12.0    4.0
4          15.0    3.0
5           5.0    NaN
8           7.0    NaN

In [24]:
pd.pivot_table(df, index='Parch', columns='Survived', values='PassengerId', aggfunc='count')

Survived      0      1
Parch                 
0         445.0  233.0
1          53.0   65.0
2          40.0   40.0
3           2.0    3.0
4           4.0    NaN
5           4.0    1.0
6           1.0    NaN

У пассажиров без семьи или с количеством людей в семье (родители, дети, родственники) до 2 было больше шансов на выживание.